### PR TITLE
[oneDNN v3.x]: Added support for matmul for fp32/bf16/int8 and einsum for fp32/bf16

### PR DESCRIPTION
**Pull request by bhavani-subramanian**
- This PR adds support for oneDNN v3.x in fused-matmul (inner-product), batch-matmul, MKL CBLAS matmul and einsum for fp32 and bf16.
- This PR also adds oneDNN v3.x support for QuantizedMatmul. Changes in oneDNNv3.x for INT8 are 1) Changing the scaling API, so we need to pass a scale for every tensor. 2) Bias needs to be float. If not already in float, we convert it to float.
- Support for weight caching in fused-matmul will be added in a future PR.
- This PR passes all unit tests related to matmul ops for fp32/bf16/int8 when oneDNN v3.x is enabled.
- This PR does not add or affect Eigen ops and are specific only to oneDNN kernels.
- For additional comments related to oneDNN v3.x integration, please refer to [this](https://github.com/tensorflow/tensorflow/pull/60307#issue-1665423762) comment.

### CODE REVIEW

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

File: tensorflow/core/kernels/mkl/mkl_qmatmul_op.cc [hunk 1]
```diff
@@ -266,6 +280,9 @@ class MklDnnQuantizedMatMulOp : public MklDnnMatMulOpBase<Tweight, Toutput> {
         // TF default format is IO. So in that case convert weight from IO
         // to OI for the first iteration and cache it to reuse in the
         // subsequent iterations, if the weight is constant.
+#ifndef ENABLE_ONEDNN_V3
+        // TODO(intel-tf): enable weights caching once it is enabled in the base
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_qmatmul_op.cc:284]: Nit:
```suggestion
        // TODO(intel-tf): enable weight caching once it is enabled in the base
```

File: tensorflow/core/kernels/mkl/mkl_qmatmul_op.cc [hunk 2]
```diff
@@ -436,13 +476,36 @@ class MklDnnQuantizedMatMulOp : public MklDnnMatMulOpBase<Tweight, Toutput> {
       output_scale.push_back(scale);
       params.post_op_params.push_back({"output_scale", output_scale});
     }
+#else
+      if (std::is_same<Toutput, quint8>::value) {
+        // Existing unit tests expect dst_scale to be computed using 256.0
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_qmatmul_op.cc:481]: Are these unit-tests oneDNN-specific? If so, we probably should just change the tests instead.

File: tensorflow/core/kernels/mkl/mkl_qmatmul_op.cc [hunk 3]
```diff
@@ -549,6 +612,111 @@ class MklDnnQuantizedMatMulOp : public MklDnnMatMulOpBase<Tweight, Toutput> {
...
+        out_scale = (255.0 * 127.0) /
+                    ((max_input - min_input) *
+                     std::max(std::abs(max_weight), std::abs(min_weight)));
+        for (int j = 0; j < n; ++j) {
+          int x = 0;
+          for (int i = 0; i < k; ++i) {
+            x += wt_buf[i * n + j];
+          }
+          if (std::is_same<Tbias, qint32>::value) {
+            // Starting oneDNN v3.0, bias is expected to be dequantized to
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_qmatmul_op.cc:669]: Nit: grammar
```suggestion
            // Starting with oneDNN v3.0, bias is expected to be dequantized to
```

File: tensorflow/core/kernels/mkl/mkl_qmatmul_op_test.cc [hunk 4]
```diff
@@ -12,7 +12,7 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
-#if defined(INTEL_MKL) && !defined(ENABLE_ONEDNN_V3)
+#if defined(INTEL_MKL)
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_qmatmul_op_test.cc:15]: Nit: Please also remove `&& !ENABLE_ONEDNN_V3` in the comment at the end of this file.
```
#endif  // INTEL_MKL && !ENABLE_ONEDNN_V3
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and sorry for the delay!

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h [hunk 5]
```diff
@@ -120,31 +136,66 @@ class MklDnnMatMulFwdPrimitive : public MklPrimitive {
...
+                  const_cast<float*>(post_op_param.param.data())),
+              *fwd_stream);
+        } else if (post_op_param.name == "dst_scale") {
+          context_.dst_scale_mem->set_data_handle(
+              static_cast<void*>(
+                  const_cast<float*>(post_op_param.param.data())),
+              *fwd_stream);
+        }
+      }
+    }
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h]: Refactored as suggested.

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h [hunk 6]
```diff
@@ -303,39 +404,67 @@ class MklDnnMatMulFwdPrimitive : public MklPrimitive {
...
+             *context_.dst_scale_mem }
+#endif  // ENABLE_ONEDNN_V3
+          });
+    } else {
+      context_.net_args.push_back({{DNNL_ARG_SRC, *context_.src_mem},
+                                   {DNNL_ARG_WEIGHTS, *context_.weight_mem},
+                                   {DNNL_ARG_BIAS, *context_.bias_mem},
+                                   {DNNL_ARG_SCRATCHPAD, *context_.sp_mem},
+                                   {DNNL_ARG_DST, *context_.dst_mem}});
+    }
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h]: Refactored as suggested.

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h [hunk 7]
```diff
@@ -419,10 +548,25 @@ class MklDnnMatMulFwdPrimitiveFactory : public MklPrimitiveFactory<T> {
...
+        key_creator.AddAsKey(post_op_param.name);
+        key_creator.AddAsKey(post_op_param.param[0]);
+      } else if (post_op_param.name == "wei_scale") {
+        DCHECK_EQ(post_op_param.param.size(), 1);
+        key_creator.AddAsKey(post_op_param.name);
+        key_creator.AddAsKey(post_op_param.param[0]);
+      } else if (post_op_param.name == "dst_scale") {
+        DCHECK_EQ(post_op_param.param.size(), 1);
+        key_creator.AddAsKey(post_op_param.name);
+        key_creator.AddAsKey(post_op_param.param[0]);
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h]: Refactored as suggested.

File: tensorflow/core/kernels/mkl/mkl_qmatmul_op.cc (hunk 1 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_qmatmul_op.cc]: Fixed.

File: tensorflow/core/kernels/mkl/mkl_qmatmul_op.cc (hunk 2 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_qmatmul_op.cc]: Yes. Fixed the test.

File: tensorflow/core/kernels/mkl/mkl_qmatmul_op.cc (hunk 3 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_qmatmul_op.cc]: Fixed.

File: tensorflow/core/kernels/mkl/mkl_qmatmul_op_test.cc (hunk 4 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_qmatmul_op_test.cc:15]: Fixed.

**Comment by mahmoud-abuzaina** [comment]
Thank you for the review and for the good suggestions. I have addressed the comments. I also included a bug fix in this [line](https://github.com/tensorflow/tensorflow/pull/60378/commits/8ba59847e722c692885ff1eaa49c94cfd01b121e#diff-273ed23e668174f96de6482ae9dc33fa1e59b6c7761690ad00d42dfb90c328ecR486).

**Review by penpornk — APPROVED** [review]
Thank you for the changes!

**Review by penpornk — CHANGES_REQUESTED** [review]
We got internal code check failures. We don't use `tsl::errors::*` anymore, they need to be replaced with their `absl` counterpart. Would you mind fixing this in all files? Thank you!

**Comment by bhavani-subramanian** [comment]
@penpornk I have addressed your review comments. Can you please take a look?

File: tensorflow/core/kernels/mkl/mkl_qmatmul_op.cc [hunk 8]
```diff
@@ -297,12 +314,20 @@ class MklDnnQuantizedMatMulOp : public MklDnnMatMulOpBase<Tweight, Toutput> {
...
+      this->GetBiasHandle(context, matmul_fwd_pd, bias_tensor, weight_tensor,
+                          cpu_stream, &temp_scaled_bias_tensor, &bias_data);
+#endif  // !ENABLE_ONEDNN_V3
+      // Execute inner-product
       matmul_fwd->Execute(src_data, weight_data, bias_data, dst_data,
-                          scratch_pad.Get(), cpu_stream);
+                          matmul_fwd_dims, scratch_pad.Get(), cpu_stream);
     } catch (dnnl::error& e) {
       string error_msg = tensorflow::strings::StrCat(
           "Status: ", e.status, ", message: ", string(e.message), ", in file ",
```
[the remark this answers is not in the document]
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl/mkl_qmatmul_op.cc:335]: Fixed.

File: tensorflow/core/kernels/mkl/mkl_matmul_op_fused.cc [hunk 9]
```diff
@@ -269,7 +272,7 @@ class MklFusedMatMulOp : public MklDnnMatMulOpBase<T, T> {
 
       // Execute fused matmul op.
       matmul_prim->Execute(src_data, weight_data, bias_data, dst_data,
-                           scratch_pad.Get(), cpu_stream);
+                           matmul_params, scratch_pad.Get(), cpu_stream);
     } catch (dnnl::error& e) {
       string error_msg = "Status: " + std::to_string(e.status) +
```
[the remark this answers is not in the document]
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl/mkl_matmul_op_fused.cc:283]: Updated.

File: tensorflow/core/common_runtime/mkl_layout_pass.cc [hunk 10]
```diff
@@ -1406,9 +1406,11 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
...
           Tensor tensor;
           if (!tensor.FromProto(*proto)) {
-            TF_CHECK_OK(errors::InvalidArgument(
-                "Could not construct Tensor from TensorProto in node: ",
-                node->name()));
+            TF_CHECK_OK(Status(
+                absl::StatusCode::kInvalidArgument,
+                absl::StrCat(
+                    "Could not construct Tensor from TensorProto in node: ",
+                    node->name())));
```
Comment (penpornk) [inline tensorflow/core/common_runtime/mkl_layout_pass.cc:1413]: Why can't we use `absl::InvalidArgumentError` like I had suggested? I think it's more compact. Example:
```suggestion
            TF_CHECK_OK(
                absl::InvalidArgumentError(
                    absl::StrCat(
                        "Could not construct Tensor from TensorProto in node: ",
                        node->name())));
```

Comment (bhavani-subramanian) [reply tensorflow/core/common_runtime/mkl_layout_pass.cc]: Agree, fixed it.

**Review by penpornk — APPROVED** [review]
Thank you for the changes!

**Comment by bhavani-subramanian** [comment]
@penpornk Thanks for approving this PR. Please let me know if you need anything else from my end.

**Comment by penpornk** [comment]
@bhavani-subramanian Could you please help fix [Windows Bazel failures](https://source.cloud.google.com/results/invocations/eba509ee-fed0-4215-b671-685ec6497b2a/log)? Thank you!

```
.\tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h(375): error C2059: syntax error: '#'
.\tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h(249): note: while compiling class template member function 'void tensorflow::MklDnnMatMulFwdPrimitive<T,T,T,T,T>::Setup(const tensorflow::MklDnnMatMulFwdParams &)'
        with
        [
            T=float
        ]
.\tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h(129): note: see reference to function template instantiation 'void tensorflow::MklDnnMatMulFwdPrimitive<T,T,T,T,T>::Setup(const tensorflow::MklDnnMatMulFwdParams &)' being compiled
        with
        [
            T=float
        ]
tensorflow/core/kernels/mkl/mkl_matmul_op_fused.cc(145): note: see reference to class template instantiation 'tensorflow::MklDnnMatMulFwdPrimitive<T,T,T,T,T>' being compiled
        with
        [
            T=float
        ]
tensorflow/core/kernels/mkl/mkl_matmul_op_fused.cc(61): note: while compiling class template member function 'void tensorflow::MklFusedMatMulOp<tensorflow::CPUDevice,float,true>::Compute(tensorflow::OpKernelContext *)'
tensorflow/core/kernels/mkl/mkl_matmul_op_fused.cc(341): note: see reference to class template instantiation 'tensorflow::MklFusedMatMulOp<tensorflow::CPUDevice,float,true>' being compiled
[17,997 / 20,875] Compiling tensorflow/compiler/mlir/lite/utils/tftext_utils.cc; 40s local, remote-cache ... (9 actions, 8 running)
Target //tensorflow/tools/pip_package:build_pip_package failed to build
INFO: Elapsed time: 258.481s, Critical Path: 84.48s
INFO: 17584 processes: 12330 remote cache hit, 4755 internal, 499 local.
FAILED: Build did NOT complete successfully
```

**Review by penpornk — APPROVED** [review]
Thank you for the fix! Let's see if this works.

**Pull request merged**